pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+
+		n := len(hostChains)
+		for i, hostChain := range hostChains {
+			// Create chain
+			if chain, ok := existingChains[utiliptables.Chain(hostChain)]; ok {
+				rules.WriteString(chain)
+				rules.WriteRune('\n')
+			} else {
+				rules.WriteString(fmt.Sprintf(fmtChain, hostChain))
+			}
+			// Flush chain
+			rulesEnd.WriteString(fmt.Sprintf(fmtFlush, hostChain))
+			// Sticky session
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --rcheck --seconds %d --reap -j %s\n", svcChain, name.String(), hostChain, info.stickyMaxAgeMinutes*60, hostChain))
+			}
+			// Roughly round robin statistically if we have more than one host
+			if n > 1 {
+				rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m statistic --mode random --probability %f -j %s\n", svcChain, name.String(), 1.0/float64(n-i), hostChain))
+			} else {
+				rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j %s\n", svcChain, name.String(), hostChain))
+			}
+			// proxy
+			rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --set -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), hostChain, protocol, hosts[i]))
+		}
+		// proxy
+		rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), info.portal.ip.String(), protocol, protocol, info.portal.port, svcChain))
+
+		for _, publicIP := range info.deprecatedPublicIPs {
+			rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"depreciatedPublicIP portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), publicIP, protocol, protocol, info.portal.port, svcChain))
+		}
+
+		for _, ingress := range info.loadBalancerStatus.Ingress {
+			if ingress.IP != "" {
+				rulesEnd.WriteString(fmt.Sprintf("-A %s -m comment --comment \"Ingress portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
+			}
+		}
+
+	}
+
+	// Delete chains no longer in use:
+	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
+	for _, chain := range newHostChains {
+		activeChains[chain] = true
+	}
+	for _, chain := range newServiceChains {
+		activeChains[chain] = true
+	}
+	for chain := range existingChains {
+		if !activeChains[chain] {
+			chainString := string(chain)
+			// Ignore chains that aren't from services.
+			// Short-circuit fail on chains without the shared prefix of services and general services chain,
+			// then fail more specifically.
+			if !strings.HasPrefix(chainString, "KUBE-S") || !(strings.HasPrefix(chainString, "KUBE-SVC-") || strings.HasPrefix(chainString, "KUBE-SERVICES")) {
```

Comment:
Shared prefix of KUBE-SERVICES and KUBE-SVC-*. Probably unnecessary
micro-optimization, fail on the shared part before failing on the whole of
either.

On Mon, Aug 10, 2015 at 8:44 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36701846>
> :
>
> > +
> > +	// Delete chains no longer in use:
> > +	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
> > +	for _, chain := range newHostChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for _, chain := range newServiceChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for chain := range existingChains {
> > +		if !activeChains[chain] {
> > +			chainString := string(chain)
> > +			// Ignore chains that aren't from services.
> > +			// Short-circuit fail on chains without the shared prefix of services and general services chain,
> > +			// then fail more specifically.
> > +			if !strings.HasPrefix(chainString, "KUBE-S") || !(strings.HasPrefix(chainString, "KUBE-SVC-") || strings.HasPrefix(chainString, "KUBE-SERVICES")) {
>
> what is KUBE-S that isn't KUBE-SVC ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36701846>
> .
>